components/resource_control/src/service.rs

```diff
@@ -139,6 +155,163 @@ impl ResourceManagerService {
+            match self
+                .pd_client
+                .load_global_config(RESOURCE_CONTROL_CONTROLLER_CONFIG_PATH.to_string())
+                .await
+            {
+                Ok((items, _)) => {
+                    match serde_json::from_slice::<ControllerConfig>(items[0].get_payload()) {
+                        Ok(c) => return Ok(c.request_unit),
+                        Err(e) => {
+                            error!("parse controller config failed"; "err" => ?e);
+                            let _ = GLOBAL_TIMER_HANDLE
+                                .delay(std::time::Instant::now() + RETRY_INTERVAL)
+                                .compat()
+                                .await;
+                            continue;
+                        }
+                    }
+                }
+                Err(err) => {
+                    error!("failed to controller config"; "err" => ?err);
+                    let _ = GLOBAL_TIMER_HANDLE
+                        .delay(std::time::Instant::now() + RETRY_INTERVAL)
+                        .compat()
+                        .await;
+                    continue;
+                }
+            }
+        }
+    }
+
+    // upload ru metrics periodically.
+    pub async fn upload_ru_metrics(&self) {
+        let mut last_consumption_map: HashMap<String, Consumption> = HashMap::new();
+        // load controller config firstly.
+        let config = self.load_controller_config().await.unwrap_or_default();
+
+        loop {
+            let background_groups: Vec<_> = self
+                .manager
+                .resource_groups
+                .iter()
+                .filter_map(|kv| {
+                    let g = kv.value();
+                    g.limiter.as_ref().map(|limiter| GroupStats {
+                        name: g.group.name.clone(),
+                        ru_quota: g.get_ru_quota() as f64,
+                        limiter: limiter.clone(),
+                        stats_per_sec: GroupStatistics::default(),
+                        expect_cost_rate: 0.0,
+                    })
+                })
+                .collect();
+            if background_groups.is_empty() {
+                let _ = GLOBAL_TIMER_HANDLE
+                    .delay(std::time::Instant::now() + BACKGROUND_RU_UPLOAD_DURATION)
+                    .compat()
+                    .await;
+                continue;
+            }
+
+            let mut req = TokenBucketsRequest::default();
+            let all_reqs = req.mut_requests();
+            background_groups.iter().for_each(|g| {
+                let mut req = TokenBucketRequest::default();
+                req.set_resource_group_name(g.name.clone());
+                req.set_is_background(true);
+                // update statistics.
+                let report_consumption = req.mut_consumption_since_last_request();
+                if let Some(last_consumption) = last_consumption_map.get_mut(g.name.as_str()) {
```

Comment:
I suggest cache the GroupStatistics here and calculate the delta value first then you needn't do any `(cur - last).max(0)` in the following code. And you can easily check if the delta value is zero, then you needn't report this request.

Reply:
done.